gluten-core/src/main/java/io/glutenproject/substrait/expression/DecimalLiteralNode.java

```diff
@@ -52,7 +52,8 @@ public Expression toProtobuf() {
   private static final byte minus_one = -1;
 
   public static byte[] encodeDecimalIntoBytes(BigDecimal decimal, int scale, int byteWidth) {
-    BigDecimal scaledDecimal = decimal.multiply(powerOfTen(scale));
+    BigDecimal scaledDecimal = BigDecimal.valueOf(decimal.unscaledValue().longValue(),
```

Comment:
When the scale become negative value, [POWER_OF_10[scale]](https://github.com/oap-project/gluten/blob/main/gluten-core/src/main/java/io/glutenproject/substrait/expression/DecimalLiteralNode.java#L103) caused ArrayIndexOutOfBoundsException. So, this change could process big decimal with negative scale.

Reply:
You're right. The Spark SQL suite "round/bround" in GlutenMathFunctionsSuite could reproduce the error ArrayIndexOutOfBoundsException. However, the Substrait don't support the decimal with non-negative scale, and Velox store the scale in uint type, we'd better to do the fallback to Vanilla Spark when the scale is negative. Currently, the native validation does validate this case that input type is decimal, does we need to validate again in Gluten, such as [here](https://github.com/oap-project/gluten/blob/main/gluten-core/src/main/java/io/glutenproject/substrait/expression/ExpressionBuilder.java#L212). @rui-mo

Reply:
Tests which enable this conf should store negative scale in decimal `withSQLConf(SQLConf.LEGACY_ALLOW_NEGATIVE_SCALE_OF_DECIMAL_ENABLED.key -> "true")`
Spark's migration guide also states that in Spark 3.0, negative scale of decimal is not allowed by default.
https://spark.apache.org/docs/latest/sql-migration-guide.html
![image](https://user-images.githubusercontent.com/52736607/206906984-84da0bda-82fc-43a2-9b4d-f2f8f8314a29.png)

Reply:
With this change, can the negative scale case correctly fallback to Spark? Maybe we can validate the negative scale in native, for example, throwing exception when the scale is negative.

Reply:
The decimal type will fall back to Spark due to [native validation] (https://github.com/oap-project/velox/blob/main/velox/substrait/SubstraitParser.cpp#L122) whatever the scale is negative or positive. This change just fixed the error ArrayIndexOutOfBoundsException, which happened in the transformer before native validation.

Reply:
> 

Thanks for your reference.

Reply:
> Tests which enable this conf should store negative scale in decimal `withSQLConf(SQLConf.LEGACY_ALLOW_NEGATIVE_SCALE_OF_DECIMAL_ENABLED.key -> "true")` Spark's migration guide also states that in Spark 3.0, negative scale of decimal is not allowed by default. https://spark.apache.org/docs/latest/sql-migration-guide.html ![image](https://user-images.githubusercontent.com/52736607/206906984-84da0bda-82fc-43a2-9b4d-f2f8f8314a29.png)

Thanks for your reference.

Reply:
Because Substrait does not support negative scale, so I think that kind of Substrait plan should not be generated. How about preventing negative scale in the Scala side during plan generation? Such as throwing exception at [ExpressionBuilder](https://github.com/oap-project/gluten/blob/main/gluten-core/src/main/java/io/glutenproject/substrait/expression/ExpressionBuilder.java#L184).

Reply:
Done